### ISSUE TITLE
Bulk persist doesn't return inserted IDs

### ISSUE THREAD

**Issue by baluselva-t-s** [quarkusio/quarkus#25299]
### Description

(Bulk persist)Persist methods which takes Iterable<Entity> or Stream<Entity> returns void. There is no way to get the IDs created

`@CheckReturnValue
    default Uni<Void> persist(Iterable<Entity> entities) {
        return JpaOperations.INSTANCE.persist(entities);
    }`

` @CheckReturnValue
    default Uni<Void> persist(Stream<Entity> entities) {
        return JpaOperations.INSTANCE.persist(entities);
    }`

Can we have these methods return Multi<String> or something similar that gets us the IDs persisted

**Comment by yrodiere** [comment]
You should be able to retrieve the IDs from the entities you passed to that method; Hibernate Reactive will set the IDs on these entities when it persists them. 

I'll let @DavideD / @gavinking comment on whether changing the return type of `persist` to return the entities (or their IDs) is relevant/doable.

**Comment by DavideD** [comment]
Yes, you can get the ids after the persist doing something like:
```
        Stream<Entity> entities = ...
        Uni<List<Object>> idsUni = Panache.withTransaction( () -> Entity.persist( entities ) )
                .map( unused -> entities.map( Entity::getId ).collect( Collectors.toList() ) );
```

or
```
        Uni<List<Object>> idsUni = Panache.withTransaction( () -> Entity.persist( entities ) )
                .map( unused -> {
                    List<Object> ids = new ArrayList<>();
                    entities.iterator().forEachRemaining( e -> ids.add( e ) );
                    return ids;
                } );
```

From the Hibernate Reactive side, I need to check the code but I think it's doable.
To be fair, it's always surprised me that the persist methods return void. But usually I need the whole entity, not just the id.

@gavinking Any opinion about this? I guess it could make the life of users easier.

**Comment by FroMage** [comment]
Pretty sure the same limitation exists in the ORM-Panache API: it doesn't return IDs. This would not be very useful to people who persist entities with non-automatic IDs, but I guess it wouldn't hurt, besides perf a little.

**Comment by loicmathieu** [comment]
> To be fair, it's always surprised me that the persist methods return void. But usually I need the whole entity, not just the id.

@DavideD this has come up a little number of times since the creation of Hibernate with Panache.

At first, there was no reactive impl so we can wrote something like this:

```java
public Response create(Product product){
    product.persist();
    return Response.created(URI.create("/products/" + product.id).build();
}
```

As after persist the object on which we call the method is updated with the id, there was no need to return it from the method. The same has been done for reactive to mimic the API but here no having the entity returned is less convenient.
We already discussed this @FroMage updating persist to return Uni<Entity> instead of Uni<Void> months and didn't decided on either.

**Comment by gavinking** [comment]
I mean, yeah, I realize that the old legacy `Session.save()` method (now deprecated in H6) returns the id. But JPA's `persist()` doesn't, so it's consistent that `persistAll()` doesn't either.

And I'm not sure how helpful it would really be: the return type would be `Uni<List<Object>>` so you would still have to typecast down to `Long` or whatever, which is not much easier than just iterating over your entities to get the ids. And, indeed, the second option is more typesafe.

**Comment by gavinking** [comment]
(Now, I mean, if Panache could expose this in a *typesafe* way, and return a `Uni<List<Long>>` or whatever, then, well, perhaps that would be useful.)

**Comment by DavideD** [comment]
> As after persist the object on which we call the method is updated with the id, there was no need to return it from the method. The same has been done for reactive to mimic the API but here no having the entity returned is less convenient.

I guess one could save a line with:
```
public Response create(Product product){
    return Response.created(URI.create("/products/" + product.persist().id).build();
}
```
But I don't think it's a big deal.

Just to give an example to people that are not familiar with the reactive code,
with Panache, the code usually looks something like:
```
    @POST
    public Uni<Response> create(Fruit fruit) {
        return Panache.withTransaction(fruit::persist)
                    .replaceWith(Response.ok(fruit).status(CREATED)::build);
    }
```

It doesn't look too bad to me.

Returning the entity is probably helpful only when you want to create a new object during the persist:
```
    @POST
    public Uni<Response> createFruit(String name) {
        return Panache.withTransaction(() -> Fruit.persist(new Fruit(name) ) )
                    .replaceWith(fruit -> Response.ok(fruit).status(CREATED).build());
    }
```

instead of
```
    @POST
    public Uni<Response> createFruit(String name) {
        Fruit fruit = new Fruit(name);
        return Panache.withTransaction(fruit::persist)
                    .replaceWith(Response.ok(fruit).status(CREATED)::build);
    }
```

I think the existing approach makes more sense.

**Comment by DavideD** [comment]
> (Now, I mean, if Panache could expose this in a typesafe way, and return a Uni<List<Long>> or whatever, then, well, perhaps that would be useful.)

Maybe we can have a sort of map function to create a projection from the created entity:
```
Uni<List<Integer>> ids = fruit.persist(f -> f.id, fruits);
```

I guess this would be similar to handle a post-persist event.
